noxfile.py

```diff
@@ -1,9 +1,47 @@
+from pathlib import Path
+
 import nox
 
-@nox.session(python=["3.11"]) # The python version should match the readthedocs configuration.
+requirements_files = list(
+    {path.name.replace(".in", "") for path in Path(".").glob("*in")}
+)
+
+
+@nox.session(name="pip-compile", python=["3.11"])
+@nox.parametrize(["req"], requirements_files, requirements_files)
```

Comment:
The first one is `arg_values_list` (i.e., the arguments that are passed to the function) and the second is `ids` (i.e, the friendly id that is used in the session name).

Reply:
so it's better to pass them through kwargs then, I suppose

Reply:
@webknjaz Hey, I'd like to confirm my understanding here before I push another commit. When you say pass the var through kwargs is this what you mean?

``@nox.parametrize("req", requirements_files)``

I guess this isn't explicitly a keyword argument but when I try that I get errors such as `argument after ** must be a mapping, not list` which seems to make sense. I'm probably missing something obvious here but I've been reading through docs and experimenting but not really getting too far. I would also prefer not to deviate too much from what @gotmax23 put together for the docs repo so unless you have a strong opinion then maybe we can leave it as-is? Thanks.

Reply:
@oraNod

kwargs are keyword arguments. And this term/syntax is used in both function definition and calling. I'm talking about the calling part.
When you have a function `def f(a): ...`, you can call it as either `f(0)`, or `f(a=0)`. The former passes the arg as positional, while the latter names the key explicitly. That's the kwarg in this context, regardless of how the function is defined.

https://realpython.com/python-kwargs-and-args/#using-the-python-kwargs-variable-in-function-definitions

Now, if you were to use the double asterisk syntax, that's a *mapping* unpacking. It can only work with data in vars that look like `dict`. Back, to the previous example, if you were to have a variable `args = (0, )` and called `f(*args)`, that would be the exact equivalent of `f(0, )` (or simply `f(0)`) — the arguments are extracted from a sequence and are passed to the function sequentially as *positional*. And if you were to have a dict like `kwargs = {'a': 0}`, and then call `f(**kwargs)` — that's passing the arguments by their exact names (it's an equivalent of `f(a=0)`). You cannot use `**` with sequences/lists, only mappings/dicts. In you case, `requirements_files` is a sequence so Python won't let you use a double asterisk syntax that exists for mappings exclusively, hence the error.

So basically, when you call functions, there's a few ways of *passing* arguments there, regardless of how they are defined in function definitions (well, technically, the definitions may apply some restrictions, but that's not important in this context).

When those `*args` and `**kwargs` syntaxes are used in function definitions, the conversion is opposite and may depend on how the call happens. For example, if you have `def f(a1, *args, **kwargs): ...` and call it like `f(a1=3, x='y')`, then internally in the function, the vars would be assigned as `a1=3`, `args=()` and `kwargs={'x': 'y'}`. But if it's called like `f(3, 5, 6, x='y')` then it'd be `a1=3`, `args=(5, 6)` and `kwargs={'x': 'y'}`. `f(1, 2, 3)` would give you `a1=1`, `args=(2, 3)` and `kwargs={}`.

So there are different ways of passing arguments. But some make it less obvious what's going on. Whenever it's obvious that a function accepts a list of numbers, it's probably fine to just enumerate them, but when a function accepts many `bool`s, and you start calling it as `f(True, True, False, True, None)`, that makes it impossible for the reader to guess what you're passing inside without having to inspect that called function in question. It also makes it possible to accidentally confuse/swap things you want to pass as you write it, without ever noticing: `f(True, False, True, True, None)`. Now, with a simple change of calling out what the arguments *are* it makes it so the cognitive complexity becomes bearable for the reader who didn't previously read and memorize the argument order and semantics of the called function prior to reading the code: `f(with_grace=True, kill_if_stuck=True, suppress_critical_errors=False, make_coffee=True, fallback_callback=None)`. Now, suddenly, there's context to the whole thing.

I'm thinking along the lines of what Brandon Rhodes was trying to explain all those years, over a decade ago: https://youtu.be/x-kB2o8sd5c?si=qvARg0O9RbVY_IyP&t=1038.
And that's basically my motivation here.

Seeing something like `@nox.parametrize(["req"], names=requirements_files, things=requirements_files)` would greatly improve the perception of what's going on there. And I wouldn't question if you passed the same variable twice by mistake as it looked to be the case initially, before Max added some context...

Reply:
I personally prefer the current form, but to adopt @webknjaz's suggestion (which is also fine with me), you'd want something like this:

```suggestion
@nox.parametrize(["req"], arg_values_list=requirements_files, ids=requirements_files)
```

Reply:
thanks a mill for the detailed explanation @webknjaz I think I just misinterpreted your initial comment but it's clear now. really appreciate you taking the time to provide all that info!

Reply:
@gotmax23 cheers. yes, that's clear to me now. my own preference here is to adopt Sviat's suggestion. as he pointed out having the keywords makes it easier to see what is going on there. I, too, had a similar question until I read through nox docs so maybe the keywords could help others avoid the need to do that. thanks for adding the suggestion.

Reply:
You're welcome :) I suppose I could've added more context in the initial comment right away. Sometimes, it's hard to guess if the receiving end would interpret the information in the same context :man_shrugging: